### PR TITLE
Milestone 3: Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by Gardimy**
In this project, I add validations to my models and create specs for them.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  it 'is invalid with a negative posts_counter' do
+    user = User.new(name: 'John', posts_counter: -1)
+    expect(user).to_not be_valid
+  end
+
+  it 'is invalid with a non-integer posts_counter' do
+    user = User.new(name: 'John', posts_counter: 1.5)
+    expect(user).to_not be_valid
+  end
```
Comment (VICTOR-CHUKWU) [inline spec/models/user_spec.rb:22]: - [ ] kindly make sure to add test for all methods in the model, currently, there is no tests for the method `recent_posts`, kindly fix that

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  it 'is invalid with a negative comments_counter' do
+    user = User.create(name: 'John', posts_counter: 0)
+    post = Post.new(title: 'Test Post', comments_counter: -1, likes_counter: 0, author: user)
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a non-integer likes_counter' do
+    user = User.create(name: 'John', posts_counter: 0)
+    post = Post.new(title: 'Test Post', comments_counter: 0, likes_counter: 1.5, author: user)
+    expect(post).to_not be_valid
```
Comment (VICTOR-CHUKWU) [inline spec/models/post_spec.rb:29]: - [ ] kindly make sure to add test for all methods in the model, currently, there are no tests for the method `update_post_counter`, kindly fix that

File: spec/models/comment_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    post = Post.create(title: 'Test Post', comments_counter: 0, likes_counter: 0, author: user)
+    comment = Comment.new(text: 'A' * 1001, post:, author: user)
+    expect(comment).to_not be_valid
+  end
+
+  it 'is invalid without a post_id' do
+    user = User.create(name: 'John', posts_counter: 0)
+    comment = Comment.new(text: 'Test comment', author: user)
+    expect(comment).to_not be_valid
+  end
```
Comment (VICTOR-CHUKWU) [inline spec/models/comment_spec.rb:29]: - [ ] kindly add a test to prove the private method `update_comments_counter` works, you can do that by calling the `after_save method`

File: spec/models/like_spec.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    post = Post.create(title: 'Test Post', comments_counter: 0, likes_counter: 0, author: user)
+    like = Like.new(post:, author: user)
+    expect(like).to be_valid
+  end
+
+  it 'is invalid without a post_id' do
+    user = User.create(name: 'John', posts_counter: 0)
+    like = Like.new(author: user)
+    expect(like).to_not be_valid
+  end
```
Comment (VICTOR-CHUKWU) [inline spec/models/like_spec.rb:15]: - [ ] kindly add a test to prove the private method `increment_like_counter` works, you can do that by calling the `after_save method`

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @Gardimy  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- model properly validated


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      # Create some posts for the user with different timestamps
+      user.posts.create(title: 'Post 1', text: 'Content 1', created_at: 2.days.ago)
+      user.posts.create(title: 'Post 2', text: 'Content 2', created_at: 1.day.ago)
+      post3 = user.posts.create(title: 'Post 3', text: 'Content 3', created_at: Time.now)
+
+      # Call the recent_posts method with a limit of 2
+      recent_posts = user.recent_posts(2)
+
+      # Expect the method to return the most recent posts (limited to 2)
+      expect(recent_posts).to eq([post3])
```
Comment (bienvenuushindi) [inline spec/models/user_spec.rb]: :red_circle: [REQUIRED] The test case (User#recent_posts returns the most recent posts) is failing :arrow_down: 
failing
As you can see the recent_posts method returns two elements but you expect one element this is the reason why this is failing. You can fix this using the code below :arrow_down: 
```
  # Expect the method to return the most recent posts (limited to 2)
      expect(recent_posts.length).to eq(2)
```

**Review by bienvenuushindi — CHANGES_REQUESTED** [review]
Hi, @Gardimy  ,

Good job so far coming so far with the progress you've made! 💪💪However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights💡
- Good commit message.✔️ 
- Good PR Summary.✔️
- The repo looks professional.:heavy_check_mark:
- There are no linter errors:heavy_check_mark:




# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Please, remember to tag me in your question so I can receive the notification.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @bienvenuushindi in your question so I can receive the notification.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by franclobo — APPROVED** [review]
Hi @Gardimy, 

# STATUS: APPROVED 😃 

git-merge


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- No linter errors ✔️ 
- Descriptive PR ✔️ 
- Correct Git flow ✔️ 
- All tests passed ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**